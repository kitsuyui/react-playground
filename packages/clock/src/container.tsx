import React from 'react'
import { useState } from 'react'
import { useInterval } from 'react-use'

import type { ClockContainerProps } from './types'

export const DateContext = React.createContext(new Date())

export const ClockContainer: React.FC<ClockContainerProps> = (
  props
): React.JSX.Element => {
  const { children } = props
  const [date, setDate] = useState(new Date())
  const refreshInterval = props.refreshInterval || 1000 // default 1 second

  // use useInterval instead of setInterval
  // https://github.com/streamich/react-use/blob/master/docs/useInterval.md
  useInterval(() => {
    setDate(new Date())
  }, refreshInterval)

  return <DateContext.Provider value={date}>{children}</DateContext.Provider>
}
