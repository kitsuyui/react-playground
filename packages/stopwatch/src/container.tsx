import React from 'react'
import { useState } from 'react'
import { useInterval } from 'react-use'

import type { StopwatchContainerProps } from './types'
export type * from './types'

// TODO: Support leap second: https://github.com/kitsuyui/react-playground/issues/40
export const StopwatchContext = React.createContext({
  elapsedTime: 0,
  running: false,
  start() {
    /* do nothing */
  },
  stop() {
    /* do nothing */
  },
  toggle() {
    /* do nothing */
  },
  reset() {
    /* do nothing */
  },
})

export function calcTimeDiff(startTime: Date, endTime: Date) {
  return (endTime.getTime() - startTime.getTime()) / 1000
}

function calcElapsedTime(startTime: Date) {
  return calcTimeDiff(startTime, new Date())
}

export const StopwatchContainer: React.FunctionComponent<
  StopwatchContainerProps
> = (props): JSX.Element => {
  const { children } = props
  const [running, setRunning] = useState(false)
  const [elapsedTimeInLap, setElapsedTimeInLap] = useState(0)
  const [elapsedTimeTotal, setElapsedTimeTotal] = useState(0)
  const [startTime, setStartTime] = useState(new Date())

  const refreshInterval = props.refreshInterval || 10 // default 10ms

  useInterval(() => {
    if (running) {
      updateElapsedTime()
    }
  }, refreshInterval)

  function updateElapsedTime() {
    setElapsedTimeInLap(calcElapsedTime(startTime))
  }

  function moveLapToTotal() {
    setElapsedTimeTotal(elapsedTimeTotal + elapsedTimeInLap)
    setElapsedTimeInLap(0)
  }

  function start() {
    if (running) {
      return
    }
    setStartTime(new Date())
    setRunning(true)
  }

  function stop() {
    if (!running) {
      return
    }
    updateElapsedTime()
    setRunning(false)
    moveLapToTotal()
  }

  function toggle() {
    if (running) {
      stop()
    } else {
      start()
    }
  }

  function reset() {
    if (running) {
      stop()
    }
    setElapsedTimeInLap(0)
    setElapsedTimeTotal(0)
  }

  const elapsedTime = elapsedTimeInLap + elapsedTimeTotal

  return (
    <StopwatchContext.Provider
      value={{
        elapsedTime,
        running,
        toggle,
        reset,
        start,
        stop,
      }}
    >
      {children}
    </StopwatchContext.Provider>
  )
}
