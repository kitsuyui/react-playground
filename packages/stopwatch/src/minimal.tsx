import React from 'react'

import { zeroPad2, zeroPad3 } from './utils'

import type { StopwatchProps } from './types'
export type * from './types'
export * from './container'

export const toLabel = (value: number) => {
  const minutes = Math.floor(value / 60)
  const seconds = value % 60 | 0
  const milliseconds = ((value % 1) * 1000) | 0
  return `${zeroPad2(minutes)}:${zeroPad2(seconds)}.${zeroPad3(milliseconds)}`
}

export const MinimalStopwatch: React.FC<StopwatchProps> = (
  props
): JSX.Element => {
  const { elapsedTime, running, toggle, reset } = props

  return (
    <div
      style={{
        width: '6em',
        fontFamily: 'monospace',
      }}
    >
      <form>
        <span>{toLabel(elapsedTime)}</span>
        <br />
        <button
          type="submit"
          name="reset"
          onClick={(e) => {
            e.preventDefault()
            reset()
          }}
        >
          Reset
        </button>
        <button
          type="submit"
          onClick={(e) => {
            e.preventDefault()
            toggle()
          }}
        >
          {running ? 'Stop' : 'Start'}
        </button>
      </form>
    </div>
  )
}
