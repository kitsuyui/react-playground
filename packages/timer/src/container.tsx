import React from 'react'
import { useState } from 'react'
import { useInterval } from 'react-use'

import type { TimerContainerProps } from './types'
export type * from './types'

// TODO: Support leap second: https://github.com/kitsuyui/react-playground/issues/40
export const TimerContext = React.createContext({
  remaining: 0,
  running: false,
  start() {
    /* do nothing */
  },
  stop() {
    /* do nothing */
  },
  toggle() {
    /* do nothing */
  },
  reset() {
    /* do nothing */
  },
  incrementTimerValue(_value: number) {
    /* do nothing */
  },
  setTimerValue(_value: number) {
    /* do nothing */
  },
})

export function calcRemaining(targetDate: Date) {
  return (targetDate.getTime() - new Date().getTime()) / 1000
}

export const TimerContainer: React.FunctionComponent<TimerContainerProps> = (
  props
): JSX.Element => {
  const { children } = props
  const emptyFn = (_event: CustomEvent) => {
    /* do nothing */
  }
  const onStart = props.onStart ?? emptyFn
  const onStop = props.onStop ?? emptyFn
  const onComplete = props.onComplete ?? emptyFn
  const onReset = props.onReset ?? emptyFn
  const [running, setRunning] = useState(false)
  const [targetDate, setTargetDate] = useState(new Date())
  const [remaining, setRemaining] = useState(0)

  const refreshInterval = props.refreshInterval || 10 // default 10ms

  useInterval(() => {
    tick()
  }, refreshInterval)

  function tick() {
    // It can't be ticked if it is not running.
    if (running) {
      const remaining = calcRemaining(targetDate)
      setRemaining(remaining)
      if (remaining <= 0) {
        setRemaining(0)
        setRunning(false)
        onComplete(new CustomEvent('complete', {}))
      }
    }
  }

  function toggle() {
    if (running) {
      stop()
    } else {
      start()
    }
  }

  function reset() {
    // First, stop the timer if it is running.
    if (running) {
      stop()
    }
    setRemaining(0)
    onReset(new CustomEvent('reset', {}))
  }

  function start() {
    // It can't be started if the remaining time is 0.
    if (remaining <= 0) {
      return
    }
    // It can't be started if it is already running.
    if (running) {
      return
    }
    setTargetDate(new Date(Date.now() + remaining * 1000))
    setRunning(true)
    onStart(new CustomEvent('start', {}))
  }

  function stop() {
    // It can't be stopped if it is not running.
    if (!running) {
      return
    }
    setRunning(false)
    onStop(new CustomEvent('stop', {}))
  }

  return (
    <TimerContext.Provider
      value={{
        remaining,
        running,
        toggle,
        reset,
        start,
        stop,
        setTimerValue(value: number) {
          stop()
          setRemaining(value || 0)
        },
        incrementTimerValue(value: number) {
          stop()
          setRemaining(remaining + value)
        },
      }}
    >
      {children}
    </TimerContext.Provider>
  )
}
