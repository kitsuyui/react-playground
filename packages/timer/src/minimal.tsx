import React from 'react'

import type { TimerProps } from './types'
export type * from './types'
export * from './container'

export function toLabel(value: number) {
  const twoDigit = new Intl.NumberFormat('en-US', {
    minimumIntegerDigits: 2,
  })
  const threeDigit = new Intl.NumberFormat('en-US', {
    minimumIntegerDigits: 3,
  })

  const minutes = Math.floor(value / 60)
  const seconds = value % 60 | 0
  const milliseconds = (value % 1) * 1000
  return `${twoDigit.format(minutes)}:${twoDigit.format(
    seconds
  )}.${threeDigit.format(milliseconds)}`
}

export const MinimalTimer: React.FunctionComponent<TimerProps> = (
  props
): JSX.Element => {
  const { remaining, running, incrementTimerValue, toggle, reset } = props

  return (
    <div
      style={{
        border: '1px solid black',
        width: '6em',
      }}
    >
      <form>
        <span>{toLabel(remaining)}</span>
        <br />
        <button
          type="submit"
          name="incrementMinute"
          onClick={(e) => {
            e.preventDefault()
            incrementTimerValue(60)
          }}
        >
          +分
        </button>
        <button
          type="submit"
          name="incrementSecond"
          onClick={(e) => {
            e.preventDefault()
            incrementTimerValue(1)
          }}
        >
          +秒
        </button>
        <br />
        <button
          type="submit"
          name="reset"
          onClick={(e) => {
            e.preventDefault()
            reset()
          }}
        >
          Reset
        </button>
        <button
          type="submit"
          onClick={(e) => {
            e.preventDefault()
            toggle()
          }}
        >
          {running ? 'Stop' : 'Start'}
        </button>
      </form>
    </div>
  )
}
