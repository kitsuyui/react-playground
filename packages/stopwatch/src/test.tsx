import { render } from '@testing-library/react'
import React from 'react'

import { calcTimeDiff } from './container'
import { toLabel } from './minimal'

import { MinimalStopwatch } from '.'

test('render MinimalStopwatch', () => {
  const props = {
    remaining: 1000,
    running: false,
    toggle: jest.fn(),
    reset: jest.fn(),
    start: jest.fn(),
    stop: jest.fn(),
  }
  const { asFragment } = render(<MinimalStopwatch elapsedTime={0} {...props} />)
  expect(asFragment()).toMatchSnapshot()
})

test('getTimeDiff returns diff in seconds', () => {
  const startTime = new Date('2021-01-01T00:00:00Z')
  const endTime = new Date('2021-01-01T00:00:01Z')
  expect(calcTimeDiff(startTime, endTime)).toBe(1)
})

test('toLabel returns formatted string', () => {
  expect(toLabel(0)).toBe('00:00.000')
  expect(toLabel(1)).toBe('00:01.000')
  expect(toLabel(59)).toBe('00:59.000')
  expect(toLabel(60)).toBe('01:00.000')
  expect(toLabel(61)).toBe('01:01.000')
  expect(toLabel(3599)).toBe('59:59.000')
})
