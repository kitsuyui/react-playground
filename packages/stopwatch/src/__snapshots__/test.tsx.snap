// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render MinimalStopwatch 1`] = `
<DocumentFragment>
  <div
    style="border: 1px solid black; width: 6em;"
  >
    <form>
      <span>
        NaN:NaN
      </span>
      <br />
      <button
        name="reset"
        type="submit"
      >
        Reset
      </button>
      <button
        type="submit"
      >
        Start
      </button>
    </form>
  </div>
</DocumentFragment>
`;
